### PR TITLE
Problem (Fix #1979): ratchet tree mixed public and private informations

### LINKED ISSUE #1979: Problem: ratchet tree mixed public and private informations
Separate private keys from the public tree, reference to https://github.com/cisco/go-mls/pull/71

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- separated `TreeSecret`

### CODE REVIEW

File: chain-tx-enclave-next/mls/src/tree.rs [hunk 1]
```diff
@@ -595,91 +532,165 @@ impl Tree {
...
-                            .expect("decrypt failed");
-                        return Some((ancestor, SecretVec::new(secret)));
-                    }
-                }
-                panic!("don't find private key to decrypt secret");
+                let pos = NodeSize::from(tree.my_pos)
+                    .direct_path(leaf_len)
+                    .iter()
+                    .position(|&p| p == ancestor)
+                    .expect("ancestor in direct path");
```
Comment (tomtau) [inline chain-tx-enclave-next/mls/src/tree.rs:677]: I guess this is to be solved in https://github.com/crypto-com/chain/issues/1784 (as one may get a malformed commit)

File: chain-tx-enclave-next/mls/src/tree.rs [hunk 2]
```diff
@@ -595,91 +532,165 @@ impl Tree {
...
+                let secret = match pos.checked_sub(1) {
```
Comment (tomtau) [inline chain-tx-enclave-next/mls/src/tree.rs:678]: maybe some comment why `pos.checked_sub(1)` and what those two cases represent (no ancestors for None?)?

File: chain-tx-enclave-next/mls/src/tree.rs [hunk 3]
```diff
@@ -689,89 +700,61 @@ impl Tree {
     /// find the overlap parent node, decrypt the secret and implant it
     ///
     /// returns (leaf_parent_hash, commit_secret)
-    fn apply_path_secrets(
+    pub fn apply_path_secrets(
```
Comment (tomtau) [inline chain-tx-enclave-next/mls/src/tree.rs:703]: or `pub (crate)` ? with some many public APIs, it's hard to know which ones are to be called from outside + what checks are expected to be performed beforehand on arguments

File: chain-tx-enclave-next/mls/src/tree.rs (hunk 1 above)
Comment (yihuang) [reply chain-tx-enclave-next/mls/src/tree.rs:677]: The ancestor is computed from `my_pos` in previous step, so by definition it should appear in `direct_path` of `my_pos`, unless there's internal bugs.

File: chain-tx-enclave-next/mls/Cargo.toml [hunk 4]
```diff
@@ -13,13 +13,14 @@ rustls = "0.18"
...
 sha2 = "0.9"
-hkdf = "0.9"
+hkdf = { version = "0.9", features = ["std"] }
 hpke = { version = "0.1.8", default-features = false, features = ["p256"] }
 aead = "0.3"
 rand = "0.7"
 chrono="0.4.13"
 ra-client = { path = "../enclave-ra/ra-client" }
 subtle = "2.2.3"
+non-empty-vec = "0.1.2"
```
Comment (tomtau) [inline chain-tx-enclave-next/mls/Cargo.toml:23]: it can use chain-util now?

Comment (yihuang) [reply chain-tx-enclave-next/mls/Cargo.toml:23]: yes, done

File: chain-tx-enclave-next/mls/src/tree.rs (hunk 3 above)
Comment (yihuang) [reply chain-tx-enclave-next/mls/src/tree.rs:703]: done

File: chain-tx-enclave-next/mls/src/tree.rs (hunk 2 above)
Comment (yihuang) [reply chain-tx-enclave-next/mls/src/tree.rs:682]: yes, added some comments now.

**Pull request merged**